spinn_front_end_common/interface/buffer_management/recorded/recorded_database.py

```diff
@@ -0,0 +1,597 @@
+
+        :param ~sqlite3.Cursor cursor:
+        :param str source_name: The global name for the source.
+            (ApplicationVertex name)
+        :param str variable_name: The name for the variable being stored
+        :param str data_type: Tyoe of data for this source_name and
+            variable_name pair
+        """
+        for row in cursor.execute(
+                """
+                SELECT data_type
+                FROM local_metadata
+                WHERE source_name = ? AND variable_name = ?
+                """, (source_name, variable_name)):
+            if row["data_type"] != data_type:
+                msg = TYPE_ERROR.format(
+                    source_name, variable_name, row["data_type"], data_type)
+                raise Exception(msg)
+
+    def _create_matrix_table(
+            self, cursor, source_name, variable_name, key, atom_ids):
+        """
+        Creates a matrix table to hold data from one core.
+
+        :param ~sqlite3.Cursor cursor:
+        :param str source_name: The global name for the source.
+            (ApplicationVertex name)
+        :param str variable_name: The name for the variable being stored
+        :param str key: name of the key column
+        :param list atom_ids: The ids for this core. Many be integers
+        :return: Name of the database table created for this data
+        """
+        table_name = self._table_name(
+            source_name, variable_name, RAW, atom_ids[0])
+        neuron_ids_str = ",".join(["'" + str(id) + "'" for id in atom_ids])
+        cursor.execute(
+            """
+            INSERT INTO local_metadata(
+                source_name, variable_name, table_name, first_neuron_id,
+                data_type)
+            VALUES(?,?,?,?,?)
+            """, (source_name, variable_name, table_name, atom_ids[0], MATRIX))
+
+        ddl_statement = "CREATE TABLE {} ({}, {})".format(
+            table_name, key, neuron_ids_str)
+        cursor.execute(ddl_statement)
+        return table_name
+
+    def _create_exists_table(
+            self, cursor, source_name, variable_name, key):
+        """
+        Creates an exists table to hold data from all cores.
+
+        :param ~sqlite3.Cursor cursor:
+        :param str source_name: The global name for the source.
+            (ApplicationVertex name)
+        :param str variable_name: The name for the variable being stored
+        :param str key: name of the key column
+        :return: Name of the database table created for this data
+        """
+        table_name = self._table_name(source_name, variable_name, RAW)
+        cursor.execute(
+            """
+            INSERT INTO local_metadata(
+                source_name, variable_name, table_name, first_neuron_id,
+                data_type)
+            VALUES(?,?,?,?,?)
+            """, (source_name, variable_name, table_name, None, EXISTS))
+
+        ddl_statement = "CREATE TABLE {} (atom_id{}, key)".format(
```

Comment:
This line looks weird

Reply:
Did you mean `"CREATE TABLE {} (atom_id, {})"` instead?

Reply:
Agreed